### ISSUE TITLE
Array-API interfaces

### ISSUE THREAD

**Issue by jorenham** [jorenham/optype#25]
https://data-apis.org/array-api/latest/API_specification/index.html

## goal

The goal is to provide `Protocols` that can be used to type e.g. [`scipy.fft.fftfreq`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.fft.fftfreq.html#scipy.fft.fftfreq) (through its `xp` kwarg) and [`scipy.fft.fftshift`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.fft.fftshift.html) (via the passed array).

## the (big) problems

- The dtype spec only defines `__eq__`, which also exists in `object`. So it's *impossible* to (structurally) type a "dtype".
- In the same way, it's impossible to (structurally) type a "device".
- Shapes are defined as tuples of "something". So the best we can do is type the number of dimensions, and ignore the axis types.

This means that static type-checkers won't be able to see the difference between a boolean array on your CPU, and a complex-floating array on a quantum computer within the core of a neutron star in the Andromeda galaxy.

## so now what?

We can still figure out the following types:

- `Array` + the dimensionality (because `shape: tuple[?, ...]` kinda tells us the `ndim`, but not the actual shape)
- `Namespace`, which is a protocollable module-type (often named `xp`), that could give us 
    - `xp.array`: the specific `Array` type (but there's still no HKT after asking for it for over 6 years...), and 
    - `xp.(bool|((u?)int|float|complex)(8|16|32|64))` -i.e. the impossible-to-type `DType` types (🎩 🪄 🐰)
    - `xp.__array_namespace_info__() -> Info` ([docs](https://data-apis.org/array-api/latest/API_specification/inspection.html)), an object with some other "magical" methods:
        - `.default_device() -> DefaultDevice`  
        - `.devices() -> list[Device]`
        - `.dtypes(*, device: Device?, kind: Kind?)` - with `Kind` a literal string category, which allows us to "categorize" the e.g. floats, ints, and complexes (assuming the lib properly typed it), i.e. as type aliases.

## problem solved: just type the untypeable

Note that we can also obtain the `Info` type via `Array`, since `Array.__array_namespace__() -> Namespace[Info]` s.t. `Namespace[InfoT].__array_namespace_info__() -> InfoT` (ignoring the versions). 
So we can get all dtype types from either `Array` or `Namespace`, and we can either one from the other, and vice-versa (i.e. bidirectionally in both ways, and also the other way around).

**Comment by jorenham** [comment]
So I too

**Comment by jorenham** [comment]
There are some "stubs" within the array-api repo already: https://github.com/data-apis/array-api/tree/2023.12/src/array_api_stubs

But they have one tiny issue; they don't make any sense at all, and are completely useless in their current state. 
Sadly, it appears to me as if it'd be faster to rewrite them from scratch, then to attempt to fix it.

For example; the use of `typing.TypeVar` is almost always incorrect. It looks as if the authors believe that a `TypeVar` can be used as some sort of global placeholder for to-be-defined types 🤷🏻.

**Comment by jorenham** [comment]
# Typing Survey of array-API `{2022,2023}.12` libraries

## Libraries

The following libraries have native support for the `2023.12` version:

- [`array_api_strict`](https://github.com/data-apis/array-api-strict)
- [`numpy`](https://github.com/numpy/numpy) (`>=2.1.0`)
- [`jax.numpy`](https://github.com/jax-ml/jax)

Through the [`array-api-compat`](https://github.com/data-apis/array-api-compat) the following libraries also support
the `2023.12` version:

- [`cupy`](https://github.com/cupy/cupy/)
- [`dask.array`](https://github.com/dask/dask)
- [`torch`](https://github.com/pytorch/pytorch)

These don't support the `2023.12` version yet, and therefore have no `__array_namespace_info__`, so require manual
analysis:

- [`sparse`](https://github.com/pydata/sparse) ([pydata/sparse#696](https://github.com/pydata/sparse/pull/696))

## Device types

The types of the `__array_namespace_info__.devices()`:

|            Package | Device Type                                                                           |
| -----------------: | ------------------------------------------------------------------------------------- |
| `array_api_strict` | [`array_api_strict._array_object.Device`][DEVICE-XP]                                  |
|             `cupy` | [`cupy.cuda.device.Device`][DEVICE-CP]                                                |
|       `dask.array` | [`array_api_compat.common._helpers._dask_device`][DEVICE-DA] <br> `\| Literal['cpu']` |
|        `jax.numpy` | [`jax.lib.xla_client.Device`][DEVICE-JNP]                                             |
|            `numpy` | `Literal['cpu']`                                                                      |
|            `torch` | [`torch.device`][DEVICE-TORCH]                                                        |
|   `sparse`(< 0.16) | `Literal['cpu']`                                                                      |

> TODO: Write as type alias (union of protocols)

## DTypes

|            Package | DType                                                           |
| -----------------: | --------------------------------------------------------------- |
| `array_api_strict` | [`array_api_strict._dtypes._DType`][DTYPE-XP]                   |
|             `cupy` | `numpy.dtype`                                                   |
|       `dask.array` | `numpy.dtype`                                                   |
|        `jax.numpy` | `numpy.dtype`                                                   |
|            `numpy` | `numpy.dtype`                                                   |
|            `torch` | [`torch.dtype`][DTYPE-TORCH]                                    |
|  `sparse` (< 0.16) | `numpy.dtype`                                                   |
|  `sparse` (≥ 0.16) | `numpy.dtype \|`[`sparse.mlir_backend._dtypes.DType`][DTYPE-IR] |

Assuming that (only) `numpy` is available, then these the above `DTypes` can be aggregated as the
following `DType` type alias.

```python
from typing import Literal, Protocol
import numpy as np

type _DTypeNP = np.dtype[np.bool | np.number[_8Bit | _16Bit | _32Bit | _64Bit]]

# https://github.com/pytorch/pytorch/blob/v2.5.0/torch/_C/__init__.pyi.in#L181-L189
class _DTypeTorch(Protocol):
    @property
    def is_floating_point(self, /) -> bool: ...
    @property
    def is_complex(self, /) -> bool: ...
    @property
    def is_signed(self, /) -> bool: ...
    @property
    def itemsize(self, /) -> Literal[1, 2, 4, 8]: ...
    def to_real(self) -> dtype: ...
    def to_complex(self) -> dtype: ...

# https://github.com/data-apis/array-api-strict/blob/2.1/array_api_strict/_dtypes.py
class _DTypeXP(Protocol):
    @property
    def _np_dtype(self, /) -> _DTypeNP: ...

# https://github.com/pydata/sparse/blob/cc3c8d9/sparse/mlir_backend/_dtypes.py
class _DTypeSparseMlir(Protocol):
    @property
    def np_dtype(self, /) -> _DTypeNP: ...

type DType = _DTypeNP | _DTypeTorch | _DTypeXP
```

It could be useful to further define specialized type aliases for the different categories, e,g,
signed/unsigned ints, floats, etc.



[DEVICE-CP]: https://docs.cupy.dev/en/stable/reference/generated/cupy.cuda.Device.html
[DEVICE-DA]: https://github.com/data-apis/array-api-compat/blob/1.9/array_api_compat/common/_helpers.py#L582-L584
[DEVICE-JNP]: https://github.com/openxla/xla/blob/4d0fe88/xla/python/xla_extension/__init__.pyi#L442-L460
[DEVICE-TORCH]: https://github.com/pytorch/pytorch/blob/v2.5.0/torch/_C/__init__.pyi.in#L106-L124
[DEVICE-XP]: https://github.com/data-apis/array-api-strict/blob/2.1/array_api_strict/_array_object.py#L46-L61

[DTYPE-IR]: https://github.com/pydata/sparse/blob/cc3c8d97c51ba28c3a823877ae7157a3475fd6b8/sparse/mlir_backend/_dtypes.py#L25-L34
[DTYPE-TORCH]: https://github.com/pytorch/pytorch/blob/v2.5.0/torch/_C/__init__.pyi.in#L181-L189
[DTYPE-XP]: https://github.com/data-apis/array-api-strict/blob/2.1/array_api_strict/_dtypes.py#L8-L39

**Comment by jorenham** [comment]
## NumPy's numeric scalar type signatures (1.24 ~ 2.1)

Distinguishable members of the numeric `numpy` scalar type kinds:

|                             | b | i | u | f | c |
|-----------------------------|---|---|---|---|---|
| `__class_getitem__()`       |   | X | X | X | X |
| `__complex__()`             | T | T | T | T | X |
| `__index__()`               | R | X | X |   |   |
| `__round__()`               |   | X | X | X | R |
| `as_integer_ratio()`        |   |   |   | X |   |
| `bit_count()`               |   | X | X |   |   |
| `denominator` / `numerator` |   | X | X |   |   |
| `is_integer()`              |   | X | X | X |   |

legend:

| <blank> | X      | R                     | T                        |
|---------|--------|-----------------------|--------------------------|
| missing | exists | runtime-only (`<2.2`) | type-check-only (`<2.2`) |

## problems

- `bool_` (b) has no unique members w.r.t. `generic` or `ndarray` when type-checking
- `int` (i) & `uint` (u) are indistinguishable
- `floating` (f) is assignable to `complexfloating` (c) when type-checking, due to the (incorrect) `floating.__complex__` annotation (fixed in `numpy>=2.2`)

## solutions

These could be the protocols:

- `Bool` (b)
  - `item() -> builtins.bool`
- `Int =: SInt | UInt` (iu)
  - `denominator`
  - `numerator`
  - `bit_count()`
- `Bool | Int` (biu)
  - ~`__index__`~ (requires `numpy >= 2.2` & deprecated)
  - `item() -> int`
- `Float` (f)
  - `as_integer_ratio()`
- `Float64`
  - `__getnewargs__() -> tuple[float]`
- `Int | Float` (iuf)
  - `is_integer()`
- `Bool | Int | Float` (biuf)
  - `item() -> float`
- ~`Complex` (c)~
  - ~`__complex__`~ (requires `numpy >= 2.2` & deprecated)
- `Complex128` (c16)
  - `__getnewargs__() -> tuple[float, float]`
- `Float64 | Complex128` (f8, c16)
  - `__getnewargs__() -> tuple[float] | tuple[float, float]`
- `Number =: Int | Float | Complex` (iufc)
  - ~`__round__`~ (requires `numpy >= 2.2`)
  - `__class_getitem__` (does that even work in protocols?)
- `Bool | Int | Float | Complex` (biufc) (like `complex_co`)
  - `item() -> complex`

**Comment by jorenham** [comment]
superseded by https://github.com/data-apis/array-api-typing
